ash/src/lib.rs

```diff
@@ -66,6 +66,40 @@ impl<'r, T> RawPtr<T> for Option<&'r T> {
     }
 }
 
+/// Given a mutable raw pointer to a type with an `s_type` member such as [`vk::BaseOutStructure`],
+/// match on a set of Vulkan structures. The struct will be rebound to the
+/// given variable of the type of the given Vulkan structure.
+///
+/// ```
+/// let mut info = ash::vk::DeviceCreateInfo::default();
+/// let info: *mut ash::vk::BaseOutStructure = <*mut _>::cast(&mut info);
+/// unsafe {
+///     ash::match_struct!(match info {
+///         info @ ash::vk::DeviceQueueCreateInfo => {
+///             dbg!(&info); // Unreachable
+///         }
+///         info @ ash::vk::DeviceCreateInfo => {
+///             dbg!(&info);
+///         }
+///     })
+/// }
+/// ```
+#[macro_export]
+macro_rules! match_struct {
```

Comment:
Naming suggestions welcome! Perhaps it should be moved into `ash/src/prelude.rs` too, but just like `mod vk` reexporting its own prelude we should surely do the same here?

(Not done currently because there's only `pub(crate)` in there)

Reply:
This seems too niche to be part of the prelude, IMO. Most downstreams of ash are not implementing layers (which seems to be the sole motivation here?).

Reply:
Let's move this to `util`, good call.

Reply:
(It was part of the library root fwiw, but that can be considered a `prelude` or at least "too easily" accessible for how niche it is).

Yes indeed, I have currently only found this to be useful to write Vulkan layers in Rust with `ash`. When the user chains structs together themselves to e.g. read properties or features, they already have access to and should use the `let` bindings for the individual structs instead of prying out their structs from the chain again.

Reply:
I thought it wasn't actually possible to export declarative macros from anywhere but the crate root?

Reply:
Right, you were replying to my question of moving to `prelude.rs`, I was convinced until writing the second comment that it actually already resided inside `prelude.rs`.

> I thought it wasn't actually possible to export declarative macros from anywhere but the crate root?

I don't actually know the rules for macros before nor after 2018. It seems that wherever they reside, the macros are only available through the crate root. As such we can move them wherever (we have an `ash/src/vk/macros.rs`, perhaps add `ash/src/macros.rs` for this non-API bit).